File: Example 18 JNA/src/example18.c
// Projeto : JNA Example #18
// Algoritmo: example18.c
// Função : : Send a union to C
// Autor(es) : Hugo Dionizio Santos
// Data : Sáb Mai 16 16:27:45 BRT 2015

#include <stdio.h>

// Seção de Variáveis Globais

// Interface do algoritmo

// Função Principal
//	Seção de Variáveis Locais
//	Seção de Comandos

// Seção de Funções
typedef union Example18Union {
	int intnumber;
	float floatnumber;
	char* stringval;
} Example18Union;

enum Example18UnionType{
    EX18_UNKNOWN = 0,
    EX18_INTEGER = 1,
    EX18_FLOAT = 2,
    EX18_STRING = 3
};

typedef struct Example18UnionHolder {
	enum Example18UnionType uniontype;
	Example18Union unionval;
} Example18UnionHolder;

void example18_sendUnion(const Example18UnionHolder* pHolder)
{
	switch (pHolder->uniontype)
	{
		case EX18_INTEGER:
		{
			printf("(C) Example 18: received integer value: %d\n", pHolder->unionval.intnumber);
			break;
		}
		case EX18_FLOAT:
		{
			printf("(C) Example 18: received float value: %f\n", pHolder->unionval.floatnumber);
			break;
		}
		case EX18_STRING:
		{
			printf("(C) Example 18: received string value: %s\n", pHolder->unionval.stringval);
			break;
		}
		default:
		{
			printf("(C) Example 18: ERROR: unrecognized type %d\n", pHolder->uniontype);
			break;
		}
	}
}

// Fim do algoritmo

File: Example 18 JNA/src/Example18.java
// Projeto : JNA Example #18
// Classe: Example18.java
// Função : : Send a union to C
// Autor(es) : Hugo Dionizio Santos (adaptações e correções), Ethan Shayne in http://www.eshayne.com/jnaex
// Data : Sáb Mai 16 16:27:45 BRT 2015

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.ptr.PointerByReference;		
import com.sun.jna.Pointer;
import com.sun.jna.Structure;
import com.sun.jna.Structure.ByReference;
import com.sun.jna.Union;

public class Example18 {
	public interface CLibrary extends Library {
    public static interface Example18UnionType {
        public static final int UNKNOWN = 0;
        public static final int INTEGER = 1;
        public static final int FLOAT = 2;
        public static final int STRING = 3;
    }    
	public static class Example18Union extends Union {
		public static class ByValue extends Example18Union implements Union.ByValue {};
		
		public int intnumber;
		public float floatnumber;
		public String stringval;
	}
	
	public static class Example18UnionHolder extends Structure {
		public static class ByReference extends Example18UnionHolder implements Structure.ByReference {}
		
		public int uniontype;
		public Example18Union unionval;
	}

	public void example18_sendUnion(Example18UnionHolder.ByReference pHolder);
	}

	static public void main(String argv[]) {
		CLibrary clib;
		CLibrary.Example18UnionHolder.ByReference ex18ref1;

		try {
			clib = (CLibrary)Native.loadLibrary("./libexample18.so", CLibrary.class);
			// 
			ex18ref1 = new CLibrary.Example18UnionHolder.ByReference();
		}
		catch (UnsatisfiedLinkError e) {
			clib = (CLibrary)Native.loadLibrary("./bin/libexample18.so", CLibrary.class);
			// 
			ex18ref1 = new CLibrary.Example18UnionHolder.ByReference();		
		}
		ex18ref1.uniontype = CLibrary.Example18UnionType.INTEGER;
		ex18ref1.unionval.setType(Integer.TYPE);
		ex18ref1.unionval.intnumber = 23;
		clib.example18_sendUnion(ex18ref1);

		CLibrary.Example18UnionHolder.ByReference ex18ref2 = new CLibrary.Example18UnionHolder.ByReference();
		ex18ref2.uniontype = CLibrary.Example18UnionType.STRING;
		ex18ref2.unionval.setType(String.class);
		ex18ref2.unionval.stringval = "hello";
		clib.example18_sendUnion(ex18ref2);

		CLibrary.Example18UnionHolder.ByReference ex18ref3 = new CLibrary.Example18UnionHolder.ByReference();
		ex18ref3.uniontype = CLibrary.Example18UnionType.FLOAT;
		ex18ref3.unionval.setType(Float.TYPE);
		ex18ref3.unionval.floatnumber = 2.3f;
		clib.example18_sendUnion(ex18ref3);
	}
}

// Fim da classe
